src/emqx_logger_jsfmt.erl

```diff
@@ -0,0 +1,225 @@
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_logger_jsfmt).
+
+-export([format/2]).
+
+-export_type([config/0]).
+
+-elvis([{elvis_style, no_nested_try_catch, #{ ignore => [emqx_logger_jsfmt]}}]).
+
+-type config() :: #{chars_limit    => pos_integer() | unlimited,
+                    depth          => pos_integer() | unlimited,
+                    legacy_header  => boolean(),
+                    max_size       => pos_integer() | unlimited,
+                    report_cb      => logger:report_cb(),
+                    single_line    => boolean()}.
+
+-define(IS_STRING(String), (is_list(String) orelse is_binary(String))).
+
+-spec format(logger:log_event(), config()) -> iodata().
+format(#{level := Level, msg := Msg, meta := Meta}, Config0) when is_map(Config0) ->
+    Config = add_default_config(Config0),
+    format(Msg, Meta#{level => Level}, Config).
+
+format(Msg, Meta, Config) ->
+    Data =
+        try format_msg(Msg, Meta, Config) of
+            MapMsg when is_map(MapMsg) ->
+                maps:merge(Meta, MapMsg);
+            Text ->
+                Meta#{msg => Text}
+        catch
+            C:R:S ->
+                Meta#{ msg => "emqx_logger_jsfmt_format_error"
+                     , fmt_raw_input => Msg
+                     , fmt_error => C
+                     , fmt_reason => R
+                     , fmt_stacktrace => S
+                     }
+        end,
+    encode(Data, Config).
+
+encode(#{error_logger := El} = Data, Config) ->
+    do_encode(Data#{error_logger := maps:without([report_cb], El)}, Config);
+encode(Data, Config) ->
+    do_encode(Data, Config).
+
+do_encode(Data0, Config) ->
+    Data = maps:without([report_cb], Data0),
+    [jiffy:encode(json_obj(Data, Config)), "\n"].
+
+format_msg({string, Chardata}, Meta, Config) ->
+    format_msg({"~ts", [Chardata]}, Meta, Config);
+format_msg({report, _} = Msg, Meta, #{report_cb := Fun} = Config)
+  when is_function(Fun,1); is_function(Fun,2) ->
+    format_msg(Msg, Meta#{report_cb => Fun}, maps:remove(report_cb, Config));
+format_msg({report, Report}, #{report_cb := Fun} = Meta, Config) when is_function(Fun, 1) ->
+    case report_cb_format(Fun, Report) of
+        true ->
+            case Fun(Report) of
+                {Format, Args} when is_list(Format), is_list(Args) ->
+                    format_msg({Format, Args}, maps:remove(report_cb, Meta), Config);
+                Other ->
+                    #{ msg => "report_cb_bad_return"
```

Comment:
I suggest changing it to '$msg' instead. This way it will appear first in the log messages, rather than in the middle.

Reply:
is that how `map()` orders fields ?

the position of each field also very much depends on the json encoder, (jiffy in our case).
not sure if it worth the effort fiddling with the ordering here.
after all, if one chose to use json format, the logs is likely to be indexed and searched web browser.